Python/Python-tokenize.c

```diff
@@ -222,10 +247,25 @@ tokenizeriter_next(tokenizeriterobject *it)
     Py_ssize_t col_offset = -1;
     Py_ssize_t end_col_offset = -1;
     if (token.start != NULL && token.start >= line_start) {
-        col_offset = _PyPegen_byte_offset_to_character_offset(line, token.start - line_start);
+        Py_ssize_t byte_offset = token.start - line_start;
+        col_offset = byte_offset - it->byte_col_offset_diff;
     }
     if (token.end != NULL && token.end >= it->tok->line_start) {
-        end_col_offset = _PyPegen_byte_offset_to_character_offset_raw(it->tok->line_start, token.end - it->tok->line_start);
+        Py_ssize_t end_byte_offset = token.end - it->tok->line_start;
+        if (lineno == end_lineno) {
+            // If the whole token is at the same line, we can just use the token.start
+            // buffer for figuring out the new column offset, since using line is not
+            // performant for very long lines.
+            Py_ssize_t token_byte_offset = token.end - token.start;
+            Py_ssize_t token_col_offset = _PyPegen_byte_offset_to_character_offset_raw(
```

Comment:
There is still a nontrivial slowdown for this call. We should be able to avoid the heap allocation of the unicode object because we already have the entire line as a unicode object in `last_line;` so we may be able to reuse it to avoid heap allocating a new unicode object per call.

Reply:
Maybe something like this?

```
static Py_ssize_t
byte_offset_to_char_offset(PyObject *unicode_obj, Py_ssize_t byte_offset)
{
    if (!PyUnicode_Check(unicode_obj)) {
        PyErr_SetString(PyExc_TypeError, "Expected a Unicode object");
        return -1;
    }

    Py_ssize_t total_length = PyUnicode_GET_LENGTH(unicode_obj);
    Py_ssize_t utf8_length;
    const char *utf8_data = PyUnicode_AsUTF8AndSize(unicode_obj, &utf8_length);

    if (utf8_data == NULL) {
        return -1;
    }

    if (byte_offset < 0 || byte_offset > utf8_length) {
        PyErr_SetString(PyExc_IndexError, "Byte offset out of range");
        return -1;
    }

    Py_ssize_t char_offset = 0;
    Py_ssize_t current_byte_offset = 0;

    while (current_byte_offset < byte_offset) {
        unsigned char c = utf8_data[current_byte_offset];
        if (c < 0x80) {
            // 1-byte character (ASCII)
            current_byte_offset += 1;
        } else if ((c & 0xE0) == 0xC0) {
            // 2-byte character
            current_byte_offset += 2;
        } else if ((c & 0xF0) == 0xE0) {
            // 3-byte character
            current_byte_offset += 3;
        } else if ((c & 0xF8) == 0xF0) {
            // 4-byte character
            current_byte_offset += 4;
        } else {
            PyErr_SetString(PyExc_ValueError, "Invalid UTF-8 sequence");
            return -1;
        }
        char_offset++;
    }

    return char_offset;
}
```

Reply:
Can the byte offset fall in the middle of a multi-byte character?

Reply:
No, because the offsets are in token boundaries